src/NumericalAlgorithms/Spectral/SwshDerivatives.cpp

```diff
@@ -0,0 +1,173 @@
+  for (const auto& mode : cached_coefficients_metadata(l_max)) {
+    // note that the libsharp transform data is stored as a pair of complex
+    // vectors: one for the transform of the real part and one for the transform
+    // of the imaginary part of the collocation data.
+    derivative_factors[mode.transform_of_real_part_offset] =
+        detail::derivative_factor<DerivativeKind>(mode.l, Spin);
+    derivative_factors[mode.transform_of_imag_part_offset] =
+        detail::derivative_factor<DerivativeKind>(mode.l, Spin);
+  }
+  // apply the sign change as appropriate due to the adjusted spin
+  // this loop is over the (complex) transform coefficients of first the real
+  // nodal data, followed by the imaginary nodal data. They receive the same
+  // derivative factors, but potentially different signs.
+  for (size_t i = 0; i < 2; i++) {
+    ComplexDiagonalModalOperator view_of_derivative_factor{
+        derivative_factors.data() +
+            i * size_of_libsharp_coefficient_vector(l_max) / 2,
+        size_of_libsharp_coefficient_vector(l_max) / 2};
+    view_of_derivative_factor *= sharp_swsh_sign_change(
+        Spin, Spin + Tags::derivative_spin_weight<DerivativeKind>, i == 0);
+  }
+  return derivative_factors;
+}
+
+template <typename DerivativeKind, int Spin>
+const ComplexDiagonalModalOperator& cached_derivative_factors(
+    const size_t l_max) noexcept {
+  const static auto lazy_derivative_operator_cache =
+      make_static_cache<CacheRange<0, swsh_derivative_maximum_l_max>>(
+          [](const size_t generator_l_max) noexcept {
+            return derivative_factors<DerivativeKind, Spin>(generator_l_max);
+          });
+  return lazy_derivative_operator_cache(l_max);
+}
+}  // namespace
+
+template <typename DerivativeKind, int Spin>
+void compute_derivative_coefficients(
+    const gsl::not_null<
+        SpinWeighted<ComplexModalVector,
+                     Spin + Tags::derivative_spin_weight<DerivativeKind>>*>
+        derivative_modes,
+    const gsl::not_null<SpinWeighted<ComplexModalVector, Spin>*>
+        pre_derivative_modes,
+    const size_t l_max, const size_t number_of_radial_points) noexcept {
+  derivative_modes->data().destructive_resize(pre_derivative_modes->size());
+  const ComplexDiagonalModalOperator& modal_derivative_operator =
+      cached_derivative_factors<DerivativeKind, Spin>(l_max);
+  // multiply each radial chunk by the appropriate derivative factors.
+  for (size_t i = 0; i < number_of_radial_points; ++i) {
+    ComplexModalVector output_mode_view{
+        derivative_modes->data().data() +
+            i * size_of_libsharp_coefficient_vector(l_max),
+        size_of_libsharp_coefficient_vector(l_max)};
+    const ComplexModalVector input_mode_view{
+        pre_derivative_modes->data().data() +
+            i * size_of_libsharp_coefficient_vector(l_max),
+        size_of_libsharp_coefficient_vector(l_max)};
+    output_mode_view = input_mode_view * modal_derivative_operator;
```

Comment:
The choice was made to have these commute, because the vector can just as easily be thought of as a 'row vector' as a 'column vector', and the diagonal operator works regardless. Besides, it would be confusing to have a non-commutative `*` operator in C++. I'll make the order reversal here, though.